firepit/welcome.js

```diff
@@ -0,0 +1,213 @@
+  const isSetup = !!firebase_bin;
+
+  if (isSetup) {
+    console.log(
+      `${chalk.green("+")} Looks like ${chalk.green("your CLI is set up")}!`
+    );
+  } else {
+    console.log(
+      `${chalk.cyan("~")} Looks like ${chalk.yellow(
+        "your CLI needs to be set up"
+      )}.\n`
+    );
+    console.log(`${chalk.cyan("~")} This may take a moment`);
+    process.stdout.write(`${chalk.green("+")} Setting up environment...`);
+
+    return new Promise(resolve => {
+      const exe_split = firebase_exe.split(" ");
+      const install = spawn(
+        exe_split[0],
+        [...exe_split.slice(1), "--tool:force-setup"],
+        {}
+      );
+
+      install.stderr.on("data", buf => {
+        readline.clearLine(process.stdout, 2);
+        readline.cursorTo(process.stdout, 2, null);
+        const line = (
+          buf
+            .toString()
+            .split("\n")
+            .filter(l => l)
+            .slice(-1)[0] || ""
+        )
+          .trim()
+          .slice(0, process.stdout.columns - 5);
+        process.stdout.write(line);
+      });
+
+      install.on("exit", () => {
+        readline.clearLine(process.stdout, 0);
+        readline.cursorTo(process.stdout, 0, null);
+        process.stdout.write(
+          `${chalk.green("+")} Alright, ${chalk.green("your CLI is set up")}!\n`
+        );
+        resolve();
+      });
+    });
+  }
+}
+
+async function GetFirebaseToolsBins() {
+  return new Promise(resolve => {
+    const exe_split = firebase_exe.split(" ");
+    const checkSpawn = spawn(exe_split[0], [
+      ...exe_split.slice(1),
+      "--tool:setup-check"
+    ]);
+    let checkSpawnData = "";
+
+    checkSpawn.stdout.on("data", buf => {
+      checkSpawnData += buf;
+    });
+
+    checkSpawn.on("close", () => {
+      firebase_bin = JSON.parse(checkSpawnData).bins[0];
+      resolve(firebase_bin);
+    });
+  });
+}
+
+async function CheckIsLoggedIn() {
+  firebase_bin = firebase_bin || (await GetFirebaseToolsBins());
+  return new Promise(resolve => {
+    const finish = SimpleSpinner(
+      `${chalk.cyan("~")} Checking your Firebase credentials`
+    );
+    const listSpawn = fork(firebase_bin, ["list", "--json"], { silent: true });
+    let listSpawnData = "";
+
+    listSpawn.stdout.on("data", buf => {
+      listSpawnData += buf;
+    });
+
+    listSpawn.on("close", () => {
+      const isLoggedIn =
+        JSON.parse(listSpawnData.toString()).status === "success";
+
+      if (!isLoggedIn) {
+        finish(
+          "🚫",
+          `${chalk.cyan(
+            "~"
+          )} Looks like you're not authenticated. ${chalk.yellow(
+            "Let's log in"
+          )}!\n`
+        );
+      } else {
+        finish("", "");
+      }
+
+      const login = fork(firebase_bin, ["login"], { stdio: "inherit" });
+      login.on("exit", () => {
+        resolve(isLoggedIn);
+      });
+    });
+  });
+}
+
+function SimpleSpinner(message) {
+  const icons = ["", ".", "..", "..."];
```

Comment:
nit: when I was using firepit the spinner looked a little jank because the length changes so my cursor-blinker was moving horizontally.

Did you consider something like `|`, `/`, `-`. `\` or another one-char sequence?